File: sdks/node/src/browser.ts
// Browser launch + lifecycle. Spawns the ShardX engine with the same
// spoofing flags the desktop launcher uses, plus pre-launch:
//
//   • resolveAutoFields    — fill timezone/language/geolocation from a
//     live geo lookup through the bound proxy.
//   • applyScreenStrategy — cap to host monitor (macOS) or replace with
//     the host monitor (Win/Linux), matching the launcher's
//     `clamp_screen_to_real_display` / `--shardx-real-screen` switch.
//   • probeUdp             — decide QUIC + WebRTC policy from a live
//     SOCKS5 UDP_ASSOCIATE probe.
import { spawn, type ChildProcess } from "node:child_process";
import { existsSync, readFileSync, rmSync, writeFileSync } from "node:fs";
import { join } from "node:path";

import { hasAutoFields, resolveAutoFields } from "./autoResolve.js";
import { geoCheckVia, type GeoInfo } from "./geo.js";
import { Profile, userDataDir, applyEngineVersion } from "./profile.js";
import { parseProxy, probeUdp, proxyToArg, type ParsedProxy } from "./proxy.js";
import type { Runtime } from "./runtime.js";
import { applyScreenStrategy, defaultScreenModeFor, type ScreenStrategy } from "./screen.js";

export type WebRtcMode   = "auto" | "block" | "tcp_only";
/** Legacy alias retained for back-compat; prefer `ScreenStrategy`. */
export type ScreenMode   = ScreenStrategy;

const noiseDefault = (): Record<string, Record<string, number | boolean>> => ({
  canvas:       { enabled: false, seed: 0 },
  webgl:        { enabled: false, seed: 0, intensity: 0 },
  audio:        { enabled: false, seed: 0 },
  client_rects: { enabled: false, seed: 0, max_offset: 0 },
  sensors:      { enabled: false, seed: 0 },
  fonts:        { enabled: false, seed: 0 },
});

/** Deterministic non-zero 32-bit FNV-1a of `<id>::<slot>`. */
function noiseSeed(id: string, slot: string): number {
  let h = 2166136261;
  const s = `${id}::${slot}`;
  for (let i = 0; i < s.length; i++) h = Math.imul(h ^ s.charCodeAt(i), 16777619);
  h >>>= 0;
  return h === 0 ? 1 : h;
}

/** Add the default noise block when absent, then fill any seed-0 vector with a
 *  stable per-profile value — without it every profile shares seed 0 and gets
 *  an identical canvas/audio/WebGL fingerprint. */
function applyNoiseSeeds(config: Record<string, unknown>, id: string): void {
  let noise = config["noise"] as Record<string, Record<string, unknown>> | undefined;
  if (!noise || typeof noise !== "object") {
    noise = noiseDefault();
    config["noise"] = noise;
  }
  for (const slot of Object.keys(noise)) {
    const block = noise[slot];
    if (block && typeof block === "object" && !block["seed"]) {
      block["seed"] = noiseSeed(id, slot);
    }
  }
}

export interface LaunchOptions {
  proxy?: string;
  cdp?: boolean;
  headless?: boolean;
  extraArgs?: string[];
  env?: Record<string, string>;
  webrtc?: WebRtcMode;
  webrtcPublicIp?: string;
  /** Override the UDP-probe auto-decision. */
  quic?: boolean;
  /** Defaults to "cap_to_host" on macOS, "use_host" on Win/Linux. */
  screenMode?: ScreenStrategy;
  probeTimeoutMs?: number;
  /** Custom user-data-dir root. Defaults to ./shardx-profiles/<id>/. */
  userDataDir?: string;
}

export class BrowserSession {
  private _stopped = false;

  constructor(
    readonly pid: number,
    readonly userDataDir: string,
    readonly cdpUrl: string | null,
    readonly process: ChildProcess,
    readonly proxyUdpMs: number | null = null,
    readonly quicEnabled: boolean = false,
    readonly webrtcMode: WebRtcMode = "auto",
    readonly geo: GeoInfo | null = null,
  ) {}

  async stop(timeoutMs = 5000): Promise<void> {
    if (this._stopped) return;
    this._stopped = true;
    if (!this.process.pid) return;
    try { this.process.kill("SIGTERM"); } catch { /* already gone */ }
    const exited = await new Promise<boolean>((resolve) => {
      const t = setTimeout(() => resolve(false), timeoutMs);
      this.process.once("exit", () => { clearTimeout(t); resolve(true); });
    });
    if (!exited) {
      try { this.process.kill("SIGKILL"); } catch { /* ignore */ }
    }
  }
}

export class Browser {
  constructor(private readonly runtime: Runtime) {}

  async launch(profile: Profile, opts: LaunchOptions = {}): Promise<BrowserSession> {
    // Auto-install on first use (high-level ShardX.launch already does
    // this; the call is here too so low-level Browser.launch users
    // don't have to remember).
    await this.runtime.install();

    const parsed: ParsedProxy | null = opts.proxy ? parseProxy(opts.proxy) : null;

    // ---- pre-launch: auto-resolve, screen strategy, UDP probe ------
    let geo: GeoInfo | null = null;
    if (hasAutoFields(profile.config)) {
      geo = await resolveAutoFields(profile.config, parsed);
    }

    const mode: ScreenStrategy = opts.screenMode ?? defaultScreenModeFor(profile.platform);
    applyScreenStrategy(profile.config, mode);

    let proxyUdpMs: number | null = null;
    if (parsed && parsed.scheme === "socks5") {
      proxyUdpMs = await probeUdp(parsed, opts.probeTimeoutMs ?? 6000);
    }
    const udpOk = proxyUdpMs !== null;
    const quicEnabled = opts.quic ?? (parsed !== null && udpOk);
    let webrtcMode: WebRtcMode = opts.webrtc ?? "auto";
    if (webrtcMode === "auto" && parsed !== null && !udpOk) webrtcMode = "tcp_only";

    // ---- profile + udd ----------------------------------------------
    const udd = userDataDir(this.runtime, profile.id, opts.userDataDir);
    console.log(`[shardx] profile '${profile.id}' → ${udd}`);
    // Keep the spoofed Chrome version coherent with the installed engine,
    // regardless of where the profile config came from (library / file / dict).
    applyEngineVersion(profile.config, this.runtime.chromiumVersion, this.runtime.greaseBrand, this.runtime.greaseVersion);
    applyNoiseSeeds(profile.config, profile.id);
    const fpFile = join(udd, "fingerprint.json");
    writeFileSync(fpFile, JSON.stringify(profile.config));

    const argv: string[] = [
      `--fingerprint-profile=${fpFile}`,
      `--user-data-dir=${udd}`,
      "--no-first-run",
    ];
    if (!profile.hasWebGPU) argv.push("--disable-features=WebGPU");
    if (!opts.headless && !opts.cdp) {
      argv.push("--restore-last-session", "--hide-crash-restore-bubble");
    }
    // Engine-side real-screen switch only fires on use_host (where the SDK
    // already rewrote screen.* — keep them in sync with the launcher).
    if (mode === "use_host") argv.push("--shardx-real-screen");
    if (parsed) {
      argv.push(`--proxy-server=${proxyToArg(parsed)}`);
      argv.push(quicEnabled ? "--enable-quic" : "--disable-quic");
    }
    if (webrtcMode === "block") {
      argv.push(
        "--force-webrtc-ip-handling-policy=disable_non_proxied_udp",
        "--shardx-webrtc-policy=block",
      );
    } else if (webrtcMode === "tcp_only") {
      argv.push(
        "--force-webrtc-ip-handling-policy=disable_non_proxied_udp",
        "--shardx-webrtc-policy=tcp_only",
      );
      // Engine spoofs the public side of ICE candidates with this IP.
      // Match the launcher: ALWAYS resolve when proxy is bound — relying
      // on `geo` from auto-resolve only works when the profile has auto
      // sentinels, otherwise the engine falls back to the host IP.
      let ip = opts.webrtcPublicIp ?? geo?.ip;
      if (!ip && parsed) {
        try { ip = (await geoCheckVia(parsed)).ip || undefined; } catch { /* leave undefined */ }
      }
      if (ip) argv.push(`--shardx-webrtc-public-ip=${ip}`);
    }
    const cdpMarker = join(udd, "DevToolsActivePort");
    if (opts.cdp) {
      if (existsSync(cdpMarker)) rmSync(cdpMarker, { force: true });
      argv.push("--remote-debugging-port=0", "--remote-allow-origins=*");
    }
    if (opts.headless) argv.push("--headless=new");
    if (opts.extraArgs) argv.push(...opts.extraArgs);

    const child = spawn(this.runtime.binaryPath, argv, {
      env: { ...process.env, ...(opts.env ?? {}) },
      stdio: "ignore",
      detached: process.platform !== "win32",
    });

    const cdpUrl = opts.cdp ? await readCdpEndpoint(udd, 15_000) : null;
    return new BrowserSession(
      child.pid!, udd, cdpUrl, child, proxyUdpMs, quicEnabled, webrtcMode, geo,
    );
  }
}

async function readCdpEndpoint(udd: string, timeoutMs: number): Promise<string | null> {
  const marker = join(udd, "DevToolsActivePort");
  const deadline = Date.now() + timeoutMs;
  while (Date.now() < deadline) {
    if (existsSync(marker)) {
      try {
        const firstLine = readFileSync(marker, "utf8").split("\n")[0].trim();
        const port = parseInt(firstLine, 10);
        if (!Number.isNaN(port)) {
          const r = await fetch(`http://127.0.0.1:${port}/json/version`);
          if (r.ok) {
            const data = await r.json() as { webSocketDebuggerUrl?: string };
            if (data.webSocketDebuggerUrl) return data.webSocketDebuggerUrl;
          }
        }
      } catch { /* keep polling */ }
    }
    await new Promise((r) => setTimeout(r, 100));
  }
  return null;
}

File: sdks/rust/src/browser.rs
//! Browser launch + lifecycle. Spawns the ShardX engine with the same
//! spoofing flags the desktop launcher uses, plus pre-launch auto-resolve,
//! screen strategy, and UDP probe.

use std::path::PathBuf;
use std::process::{Child, Command, Stdio};
use std::sync::Arc;
use std::time::Duration;

use anyhow::Result;

use crate::auto_resolve::{has_auto_fields, resolve_auto_fields};
use crate::geo::{geo_check_via, GeoInfo};
use crate::profile::{user_data_dir, Profile};
use crate::proxy::{parse_proxy, probe_udp, proxy_to_arg, ParsedProxy, ProxyScheme};
use crate::runtime::Runtime;
use crate::screen::{apply_screen_strategy, default_screen_mode_for, ScreenStrategy};

/// Deterministic non-zero 32-bit FNV-1a of `<id>::<slot>`.
fn noise_seed(id: &str, slot: &str) -> u32 {
    let mut h: u32 = 2166136261;
    for b in format!("{id}::{slot}").bytes() {
        h = (h ^ b as u32).wrapping_mul(16777619);
    }
    if h == 0 {
        1
    } else {
        h
    }
}

/// Add the default noise block when absent, then fill any seed-0 vector with a
/// stable per-profile value — without it every profile shares seed 0 and gets
/// an identical canvas/audio/WebGL fingerprint.
fn apply_noise_seeds(config: &mut serde_json::Value, id: &str) {
    let Some(obj) = config.as_object_mut() else {
        return;
    };
    obj.entry("noise").or_insert_with(|| {
        serde_json::json!({
            "canvas":       { "enabled": false, "seed": 0 },
            "webgl":        { "enabled": false, "seed": 0, "intensity": 0 },
            "audio":        { "enabled": false, "seed": 0 },
            "client_rects": { "enabled": false, "seed": 0, "max_offset": 0 },
            "sensors":      { "enabled": false, "seed": 0 },
            "fonts":        { "enabled": false, "seed": 0 }
        })
    });
    if let Some(noise) = obj.get_mut("noise").and_then(|n| n.as_object_mut()) {
        for (slot, block) in noise.iter_mut() {
            if let Some(b) = block.as_object_mut() {
                let needs = b
                    .get("seed")
                    .and_then(|v| v.as_u64())
                    .map(|n| n == 0)
                    .unwrap_or(true);
                if needs {
                    b.insert("seed".into(), serde_json::Value::from(noise_seed(id, slot)));
                }
            }
        }
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Default)]
pub enum WebRtcMode {
    #[default]
    Auto,
    Block,
    TcpOnly,
}

#[derive(Default)]
pub struct LaunchOptions {
    pub proxy: Option<String>,
    pub cdp: bool,
    pub headless: bool,
    pub extra_args: Vec<String>,
    pub env: Vec<(String, String)>,
    pub webrtc: Option<WebRtcMode>,
    pub webrtc_public_ip: Option<String>,
    /// Override the UDP-probe auto-decision for QUIC.
    pub quic: Option<bool>,
    /// Defaults to `CapToHost` on macOS, `UseHost` on Win/Linux.
    pub screen_mode: Option<ScreenStrategy>,
    pub probe_timeout_ms: Option<u64>,
    /// Custom user-data-dir root. Defaults to `<profiles_root>/<id>/`.
    pub user_data_dir: Option<PathBuf>,
    /// When picking a random profile, filter by `navigator.platform` substring.
    pub platform: Option<String>,
    /// Re-pick hardware_concurrency / device_memory / platform_version.
    pub randomize: bool,
}

/// A running engine process + the decisions made at launch.
pub struct BrowserSession {
    pub pid: u32,
    pub user_data_dir: PathBuf,
    pub cdp_url: Option<String>,
    pub proxy_udp_ms: Option<u128>,
    pub quic_enabled: bool,
    pub webrtc_mode: WebRtcMode,
    pub geo: Option<GeoInfo>,
    child: Option<Child>,
    stopped: bool,
}

impl BrowserSession {
    /// Terminate the engine process. Idempotent.
    pub async fn stop(&mut self) -> Result<()> {
        if self.stopped {
            return Ok(());
        }
        self.stopped = true;
        if let Some(mut child) = self.child.take() {
            let _ = child.kill();
            // Reap so we don't leave a zombie.
            tokio::task::spawn_blocking(move || {
                let _ = child.wait();
            })
            .await
            .ok();
        }
        Ok(())
    }
}

impl Drop for BrowserSession {
    fn drop(&mut self) {
        if !self.stopped {
            if let Some(mut child) = self.child.take() {
                let _ = child.kill();
            }
        }
    }
}

pub struct Browser {
    runtime: Arc<Runtime>,
}

impl Browser {
    pub fn new(runtime: Arc<Runtime>) -> Self {
        Self { runtime }
    }

    pub async fn launch(&self, mut profile: Profile, opts: LaunchOptions) -> Result<BrowserSession> {
        self.runtime.install(false).await?;

        let parsed: Option<ParsedProxy> = match &opts.proxy {
            Some(u) => Some(parse_proxy(u)?),
            None => None,
        };

        // ---- pre-launch: auto-resolve, screen strategy, UDP probe ----
        let mut geo: Option<GeoInfo> = None;
        if has_auto_fields(&profile.config) {
            geo = resolve_auto_fields(&mut profile.config, parsed.as_ref()).await;
        }

        let mode = opts
            .screen_mode
            .unwrap_or_else(|| default_screen_mode_for(&profile.platform()));
        apply_screen_strategy(&mut profile.config, mode);

        let mut proxy_udp_ms: Option<u128> = None;
        if let Some(p) = &parsed {
            if p.scheme == ProxyScheme::Socks5 {
                proxy_udp_ms = probe_udp(p, opts.probe_timeout_ms.unwrap_or(6000)).await.ok();
            }
        }
        let udp_ok = proxy_udp_ms.is_some();
        let quic_enabled = opts.quic.unwrap_or(parsed.is_some() && udp_ok);
        let mut webrtc_mode = opts.webrtc.unwrap_or(WebRtcMode::Auto);
        if webrtc_mode == WebRtcMode::Auto && parsed.is_some() && !udp_ok {
            webrtc_mode = WebRtcMode::TcpOnly;
        }

        // ---- profile + udd ----
        let udd = user_data_dir(&self.runtime, &profile.id, opts.user_data_dir.as_deref())?;
        eprintln!("[shardx] profile '{}' → {}", profile.id, udd.display());
        // Keep the spoofed Chrome version coherent with the installed engine,
        // regardless of where the profile config came from (library / file / value).
        let (grease_brand, grease_version) = self.runtime.grease();
        crate::profile::apply_engine_version(
            &mut profile.config,
            &self.runtime.chromium_version(),
            grease_brand.as_deref(),
            grease_version.as_deref(),
        );
        apply_noise_seeds(&mut profile.config, &profile.id);
        let fp_file = udd.join("fingerprint.json");
        std::fs::write(&fp_file, serde_json::to_string(&profile.config)?)?;

        let mut argv: Vec<String> = vec![
            format!("--fingerprint-profile={}", fp_file.display()),
            format!("--user-data-dir={}", udd.display()),
            "--no-first-run".into(),
        ];
        if !profile.has_webgpu() {
            argv.push("--disable-features=WebGPU".into());
        }
        if !opts.headless && !opts.cdp {
            argv.push("--hide-crash-restore-bubble".into());
        }
        if mode == ScreenStrategy::UseHost {
            argv.push("--shardx-real-screen".into());
        }
        if let Some(p) = &parsed {
            argv.push(format!("--proxy-server={}", proxy_to_arg(p)));
            argv.push(if quic_enabled { "--enable-quic" } else { "--disable-quic" }.into());
        }
        match webrtc_mode {
            WebRtcMode::Block => {
                argv.push("--force-webrtc-ip-handling-policy=disable_non_proxied_udp".into());
                argv.push("--shardx-webrtc-policy=block".into());
            }
            WebRtcMode::TcpOnly => {
                argv.push("--force-webrtc-ip-handling-policy=disable_non_proxied_udp".into());
                argv.push("--shardx-webrtc-policy=tcp_only".into());
                // Engine spoofs the public side of ICE candidates with this IP.
                let mut ip = opts
                    .webrtc_public_ip
                    .clone()
                    .or_else(|| geo.as_ref().map(|g| g.ip.clone()).filter(|s| !s.is_empty()));
                if ip.is_none() {
                    if let Some(p) = &parsed {
                        if let Ok(g) = geo_check_via(Some(p), "ip-api.com").await {
                            if !g.ip.is_empty() {
                                ip = Some(g.ip);
                            }
                        }
                    }
                }
                if let Some(ip) = ip {
                    argv.push(format!("--shardx-webrtc-public-ip={ip}"));
                }
            }
            WebRtcMode::Auto => {}
        }
        if opts.cdp {
            let marker = udd.join("DevToolsActivePort");
            let _ = std::fs::remove_file(&marker);
            argv.push("--remote-debugging-port=0".into());
            argv.push("--remote-allow-origins=*".into());
        }
        if opts.headless {
            argv.push("--headless=new".into());
        }
        argv.extend(opts.extra_args.iter().cloned());

        let mut cmd = Command::new(self.runtime.binary_path());
        cmd.args(&argv)
            .stdin(Stdio::null())
            .stdout(Stdio::null())
            .stderr(Stdio::null());
        for (k, v) in &opts.env {
            cmd.env(k, v);
        }
        let child = cmd.spawn()?;
        let pid = child.id();

        let cdp_url = if opts.cdp {
            read_cdp_endpoint(&udd, 15_000).await
        } else {
            None
        };

        Ok(BrowserSession {
            pid,
            user_data_dir: udd,
            cdp_url,
            proxy_udp_ms,
            quic_enabled,
            webrtc_mode,
            geo,
            child: Some(child),
            stopped: false,
        })
    }
}

async fn read_cdp_endpoint(udd: &std::path::Path, timeout_ms: u64) -> Option<String> {
    let marker = udd.join("DevToolsActivePort");
    let deadline = tokio::time::Instant::now() + Duration::from_millis(timeout_ms);
    let client = reqwest::Client::new();
    while tokio::time::Instant::now() < deadline {
        if let Ok(text) = std::fs::read_to_string(&marker) {
            if let Some(first) = text.lines().next() {
                if let Ok(port) = first.trim().parse::<u16>() {
                    if let Ok(resp) = client
                        .get(format!("http://127.0.0.1:{port}/json/version"))
                        .send()
                        .await
                    {
                        if resp.status().is_success() {
                            if let Ok(v) = resp.json::<serde_json::Value>().await {
                                if let Some(ws) =
                                    v.get("webSocketDebuggerUrl").and_then(|x| x.as_str())
                                {
                                    return Some(ws.to_string());
                                }
                            }
                        }
                    }
                }
            }
        }
        tokio::time::sleep(Duration::from_millis(100)).await;
    }
    None
}
